### PR TITLE
[LWG motion 22] P2641R4 Checking if a union alternative is active

### LINKED ISSUE #6308: [2023-06 LWG Motion 22] P2641R4 Checking if a union alternative is active

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #6308.

### CODE REVIEW

File: source/meta.tex [hunk 1]
```diff
@@ -583,6 +583,8 @@
 
   // \ref{meta.const.eval}, constant evaluation context
   constexpr bool is_constant_evaluated() noexcept;
+  template<class T>
+    consteval bool is_within_lifetime(const T*) noexcept;
```
Comment (JohelEGP) [inline source/meta.tex:587]: See #4822.
```suggestion
  consteval bool is_within_lifetime(const auto*) noexcept;
```

File: source/meta.tex [hunk 2]
```diff
@@ -2432,6 +2434,61 @@
 \end{example}
 \end{itemdescr}
 
+\indexlibraryglobal{is_within_lifetime}%
+\begin{itemdecl}
+template<class T>
+  consteval bool is_within_lifetime(const T* p) noexcept;
```
Comment (JohelEGP) [inline source/meta.tex:2440]: See #4822.
```suggestion
consteval bool is_within_lifetime(const auto* p) noexcept;
```

File: source/support.tex [hunk 3]
```diff
@@ -750,6 +750,7 @@
 #define @\defnlibxname{cpp_lib_unreachable}@                       202202L // also in \libheader{utility}
 #define @\defnlibxname{cpp_lib_unwrap_ref}@                        201811L // also in \libheader{type_traits}
 #define @\defnlibxname{cpp_lib_variant}@                           202106L // also in \libheader{variant}
+#define @\defnlibxname{cpp_lib_within_lifetime}@                   202306L // also in \libheader{type_traits}
```
Comment (JohelEGP) [inline source/support.tex:753]: Whyever did they exclude `is_`?
There's plenty of `<type_traits>` macros with it,
including that of its neighbour `is_constant_evaluated`.

**Review by JohelEGP — APPROVED** [review]
I'm using <https://isocpp.org/files/papers/P2641R4.html>.

File: source/meta.tex [hunk 4]
```diff
@@ -2432,6 +2434,61 @@
...
+  // * the value representations for \tcode{true} and \tcode{false} are distinct
+  //   from the value representation for \tcode{2}
+  constexpr OptBool() : c(2) { }
+  constexpr OptBool(bool b) : b(b) { }
+
+  constexpr auto has_value() const -> bool {
+    if consteval {
+      return std::is_within_lifetime(&b);       // during constant evaluation, cannot read from \tcode{c}
+    } else {
+      return c != 2;                            // during runtime, must read from \tcode{c}
```
Comment (frederick-vs-ja) [inline source/meta.tex:2474]: This line might have UB. Would it be better to use `std::bit_cast<char>(*this) != 2` instead?

Comment (jwakely) [reply source/meta.tex:2474]: You can always read through a char.

Comment (jwakely) [reply source/meta.tex:2474]: And the example notes that it relies on common implementation properties. It's a non-normative example, and the committee did not want to clutter it with numerous static assertions to make pedants happy. On common implementations, it works as intended. It's not supposed to be demonstrative of 100% portable code. It's supposed to show how to use `is_within_lifetime`.

Comment (jensmaurer) [reply source/meta.tex:2474]: "Reading through a char" has nothing to do with type punning via a union; it's just reading the object representation, which ought to be valid in all cases. (We have outstanding specification bugs in that area, though.)

File: source/meta.tex (hunk 1 above)
Comment (burblebee) [reply source/meta.tex:587]: Thanks, done as a separate commit.

**Review by JohelEGP — APPROVED** [review]

File: source/meta.tex (hunk 4 above)
Comment (jensmaurer) [reply source/meta.tex:2472]: As I said before, "We have outstanding specification bugs in that area, though.", so "under the current wording" is dubious.  But it's plausible that an lvalue of type "char" refers to the object representation of an object, not necessarily the designated member.

File: source/support.tex (hunk 3 above)
Comment (tkoeppe) [reply source/support.tex:784]: We can fix that.

File: source/meta.tex (hunk 4 above)
Comment (tkoeppe) [reply source/meta.tex:2472]: Please use the relevant issue trackers for subsequent discussions and clarifications (if any).

**Pull request merged**